ludwig/modules/metric_modules.py

```diff
@@ -241,13 +245,65 @@ def get_inputs(cls):
         return PREDICTIONS
 
 
-# TODO(shreya): Double check difference in computation.
 @register_metric(R2, [NUMBER, VECTOR])
-class R2Score(_R2Score, LudwigMetric):
-    """R-squared metric."""
+class R2Score(LudwigMetric):
+    """Custom R-squared metric implementation that modifies torchmetrics R-squared implementation to return Nan
+    when there is only sample. This is because R-squared is only defined for two or more samples.
+
+    Custom implementation uses code from torchmetrics v0.9.2's implementation of R2:
+    https://github.com/Lightning-AI/metrics/blob/master/src/torchmetrics/regression/r2.py
+    """
+
+    def __init__(
+        self, num_outputs: int = 1, adjusted: int = 0, multioutput: str = "uniform_average", **kwargs: Any
+    ) -> None:
+        super().__init__(**kwargs)
+
+        self.num_outputs = num_outputs
+
+        if adjusted < 0 or not isinstance(adjusted, int):
+            raise ValueError("`adjusted` parameter should be an integer larger or equal to 0.")
+        self.adjusted = adjusted
+
+        allowed_multioutput = ("raw_values", "uniform_average", "variance_weighted")
+        if multioutput not in allowed_multioutput:
+            raise ValueError(
+                f"Invalid input to argument `multioutput`. Choose one of the following: {allowed_multioutput}"
+            )
+        self.multioutput = multioutput
 
-    def __init__(self, num_outputs: int = 1, **kwargs):
-        super().__init__(num_outputs=num_outputs, dist_sync_fn=gather_all_tensors)
+        self.add_state("sum_squared_error", default=torch.zeros(self.num_outputs), dist_reduce_fx="sum")
+        self.add_state("sum_error", default=torch.zeros(self.num_outputs), dist_reduce_fx="sum")
+        self.add_state("residual", default=torch.zeros(self.num_outputs), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds: Tensor, target: Tensor) -> None:
+        """Update state with predictions and targets.
+
+        Args:
+            preds: Predictions from model
+            target: Ground truth values
+        """
+        sum_squared_error, sum_error, residual, n_obs = _r2_score_update(preds, target)
+
+        self.sum_squared_error += sum_squared_error
+        self.sum_error += sum_error
+        self.residual += residual
+        self.total += n_obs
+
+    def compute(self) -> Tensor:
+        """Computes r2 score over the metric states."""
+
+        # self.total maps to the number of observations in preds/target computed during update()
+        if self.total <= 1:
+            logger.warning(
+                """R-squared (r2) is not defined for one sample. It needs at least two samples. Returning NaN."""
+            )
+            return np.nan
```

Comment:
The function's return type is `Tensor`, but this looks like a numpy type. Should this return `torch.tensor[float('nan')]` instead?